server/src/main/java/org/opensearch/indices/replication/SegmentFileTransferHandler.java

```diff
@@ -0,0 +1,239 @@
+import org.opensearch.common.bytes.BytesArray;
+import org.opensearch.common.bytes.BytesReference;
+import org.opensearch.common.lease.Releasable;
+import org.opensearch.common.lucene.store.InputStreamIndexInput;
+import org.opensearch.common.util.CancellableThreads;
+import org.opensearch.core.internal.io.IOUtils;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.index.store.Store;
+import org.opensearch.index.store.StoreFileMetadata;
+import org.opensearch.indices.recovery.FileChunkWriter;
+import org.opensearch.indices.recovery.MultiChunkTransfer;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.RemoteTransportException;
+import org.opensearch.transport.Transports;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.function.IntSupplier;
+
+/**
+ * SegmentFileSender handles building and starting a {@link MultiChunkTransfer} to orchestrate sending chunks to a given targetNode.
+ * This class delegates to a {@link FileChunkWriter} to handle the transport of chunks.
+ *
+ * @opensearch.internal
+ * // TODO: make this package-private after combining recovery and replication into single package.
+ */
+public final class SegmentFileTransferHandler {
+
+    private final Logger logger;
+    private final IndexShard shard;
+    private final FileChunkWriter chunkWriter;
+    private final ThreadPool threadPool;
+    private final int chunkSizeInBytes;
+    private final int maxConcurrentFileChunks;
+    private final DiscoveryNode targetNode;
+    private final CancellableThreads cancellableThreads;
+
+    public SegmentFileTransferHandler(
+        IndexShard shard,
+        DiscoveryNode targetNode,
+        FileChunkWriter chunkWriter,
+        Logger logger,
+        ThreadPool threadPool,
+        CancellableThreads cancellableThreads,
+        int fileChunkSizeInBytes,
+        int maxConcurrentFileChunks
+    ) {
+        this.shard = shard;
+        this.targetNode = targetNode;
+        this.chunkWriter = chunkWriter;
+        this.logger = logger;
+        this.threadPool = threadPool;
+        this.cancellableThreads = cancellableThreads;
+        this.chunkSizeInBytes = fileChunkSizeInBytes;
+        // if the target is on an old version, it won't be able to handle out-of-order file chunks.
+        this.maxConcurrentFileChunks = maxConcurrentFileChunks;
+    }
+
+    /**
+     * Returns a closeable {@link MultiChunkTransfer} to initiate sending a list of files.
+     * Callers are responsible for starting the transfer and closing the resource.
+     * @param store {@link Store}
+     * @param files {@link StoreFileMetadata[]}
+     * @param translogOps {@link IntSupplier}
+     * @param listener {@link ActionListener}
+     * @return {@link MultiChunkTransfer}
+     */
+    public MultiChunkTransfer<StoreFileMetadata, FileChunk> sendFiles(
```

Comment:
I know that you basically moved the existing implementation (largely), but the original method did send files - this one does not. I would suggest to rename it to `prepareSendFiles` or `newTransfer` or something along this lines to reflect that.